### ISSUE TITLE
[Discuss] Byte schema compatibility issue

### ISSUE THREAD

**Issue by Technoboy-** [apache/pulsar#13701]
## Motivation

For now, the byte-schema is compatible with all other schemas. This may introduce more issues.

Case 1 :
```
Consumer<User> consumer = pulsarClient
                .newConsumer(Schema.AVRO(User.class))
                .topic("persistent://public/default/test")
                .subscriptionName("sub-2")
                .subscriptionType(SubscriptionType.Shared)
                .subscribe();
while(true){
     Message<User> message = consumer.receive();
     System.out.println("received msg : " + message.getValue());
}
```
If we produce byte array msg to the topic, it succeeds. 
```
bin/pulsar-client produce -n 5 -m "hello" test
```

But the consume may face the below error :
```
Exception in thread "main" com.google.common.util.concurrent.UncheckedExecutionException: org.apache.commons.lang3.SerializationException: Failed at fetching schema info for EMPTY
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4935)
	at org.apache.pulsar.client.impl.schema.reader.AbstractMultiVersionReader.getSchemaReader(AbstractMultiVersionReader.java:83)
	at org.apache.pulsar.client.impl.schema.reader.AbstractMultiVersionReader.read(AbstractMultiVersionReader.java:90)
	at org.apache.pulsar.client.impl.schema.AbstractStructSchema.decode(AbstractStructSchema.java:67)
	at org.apache.pulsar.client.impl.MessageImpl.decode(MessageImpl.java:470)
	at org.apache.pulsar.client.impl.MessageImpl.getValue(MessageImpl.java:448)
	at com.tboy.learn.pulsar.sql.TestConsumer.main(TestConsumer.java:24)
Caused by: org.apache.commons.lang3.SerializationException: Failed at fetching schema info for EMPTY
	at org.apache.pulsar.client.impl.schema.reader.AbstractMultiVersionReader.getSchemaInfoByVersion(AbstractMultiVersionReader.java:129)
	at org.apache.pulsar.client.impl.schema.reader.MultiVersionAvroReader.loadReader(MultiVersionAvroReader.java:47)
	at org.apache.pulsar.client.impl.schema.reader.AbstractMultiVersionReader$1.load(AbstractMultiVersionReader.java:52)
	at org.apache.pulsar.client.impl.schema.reader.AbstractMultiVersionReader$1.load(AbstractMultiVersionReader.java:49)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
```
This is because it can't find schema to decode.

Case 2 :

```
Consumer<byte[]> consumer = pulsarClient
                .newConsumer()
                .topic("persistent://public/default/test")
                .subscriptionName("sub-2")
                .subscriptionType(SubscriptionType.Shared)
                .subscribe();
while(true){
     Message<byte[]> message = consumer.receive();
     System.out.println("received msg : " + message.getValue());
}
```
In this case, if producer send with JSON object, consumer never know how to deserialize the message. 

### Goal
Byte schema should also follow the schema compatibility policy. So the above cases, the client should throw 'PulsarClientException$IncompatibleSchemaException' when init.

### Modification
- Producer should attach byte schema info when lookup. 

### Compatibility
- Only Broker upgrade, the client will keep the original behavior. Nothing changed.
- Only Client upgrade, so it will execute in above way.

**Comment by Technoboy-** [comment]
Here for track the schema issue.
@codelipenghui @gaoran10 @hangc0276

**Comment by mattisonchao** [comment]
@Technoboy-  

In this case, I think we need to check the current topic schema if fit we want to create producer/consumer.
If not, we can refuse the action.

Another idea: 

- Maybe we can support multiple schema type in one topic?

**Comment by Technoboy-** [comment]
Closed by #15712
